### Задание 1. Загрузка данных

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names), загрузить его и оставить 8 столбцов для дальнейшего изучения: `surgery?`, `Age`, `rectal temperature`, `pulse`, `respiratory rate`, `temperature of extremities`, `pain`, `outcome`. 

### Задание 2. Первичное изучение данных

Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.


### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [1]:
import pandas as pd
import numpy as np

### Задание 1. Загрузка данных

In [2]:
horse = pd.read_csv('horse_data.csv', na_values = ['?', np.nan])
sick_horse = horse.iloc[:, [0, 1, 3, 4, 5, 6, 10, 22]]
sick_horse.columns = ['surgery', 'age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']
sick_horse = sick_horse.astype({'rectal temperature':'float64'})
sick_horse.head(10)

surgery  age  rectal temperature  pulse  respiratory rate  \
0      1.0    1                39.2   88.0              20.0   
1      2.0    1                38.3   40.0              24.0   
2      1.0    9                39.1  164.0              84.0   
3      2.0    1                37.3  104.0              35.0   
4      2.0    1                 NaN    NaN               NaN   
5      1.0    1                37.9   48.0              16.0   
6      1.0    1                 NaN   60.0               NaN   
7      2.0    1                 NaN   80.0              36.0   
8      2.0    9                38.3   90.0               NaN   
9      1.0    1                38.1   66.0              12.0   

   temperature of extremities  pain  outcome  
0                         NaN   3.0      3.0  
1                         1.0   3.0      1.0  
2                         4.0   2.0      2.0  
3                         NaN   NaN      2.0  
4                         2.0   2.0      1.0  
5                         1.0   3.0      1.0  
6                         3.0   NaN      2.0  
7                         3.0   4.0      3.0  
8                         1.0   5.0      1.0  
9                         3.0   3.0      1.0

In [3]:
sick_horse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     298 non-null    float64
 1   age                         299 non-null    int64  
 2   rectal temperature          239 non-null    float64
 3   pulse                       275 non-null    float64
 4   respiratory rate            241 non-null    float64
 5   temperature of extremities  243 non-null    float64
 6   pain                        244 non-null    float64
 7   outcome                     298 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.8 KB


In [4]:
sick_horse.describe(include = 'all').transpose()

count       mean        std   min   25%   50%  \
surgery                     298.0   1.395973   0.489881   1.0   1.0   1.0   
age                         299.0   1.642140   2.177300   1.0   1.0   1.0   
rectal temperature          239.0  38.166527   0.733508  35.4  37.8  38.2   
pulse                       275.0  71.934545  28.680522  30.0  48.0  64.0   
respiratory rate            241.0  30.427386  17.678256   8.0  18.0  24.0   
temperature of extremities  243.0   2.345679   1.046369   1.0   1.0   3.0   
pain                        244.0   2.942623   1.303993   1.0   2.0   3.0   
outcome                     298.0   1.550336   0.737967   1.0   1.0   1.0   

                             75%    max  
surgery                      2.0    2.0  
age                          1.0    9.0  
rectal temperature          38.5   40.8  
pulse                       88.0  184.0  
respiratory rate            36.0   96.0  
temperature of extremities   3.0    4.0  
pain                         4.0    5.0  
outcome                      2.0    3.0

In [5]:
sick_horse.mode()

surgery  age  rectal temperature  pulse  respiratory rate  \
0      1.0    1                38.0   48.0              20.0   

   temperature of extremities  pain  outcome  
0                         3.0   3.0      1.0

### Выбросы.
Воспользуемся описанием значений столбцов.

столбец Age. Максимальное значение 9, а по описанию 2. Выбросы есть

столбец pulse. Максимальное значение 184, а по описанию 20-40. Выбросы есть

столбец respiratory rate. Максимальное значение 96, а по описанию 8-10 (хотя в описании написано, что полезность столбца сомнительна, из-за больших отклонений) Все остальные столбцы в содержат значения в допустимой норме, исходя из описания.

1: surgery? 1 = Yes, it had surgery 2 = It was treated without surgery

2: Age 1 = Adult horse 2 = Young (< 6 months)

4: rectal temperature - linear - in degrees celsius. - An elevated temp may occur due to infection. - temperature may be reduced when the animal is in late shock - normal temp is 37.8 - this parameter will usually change as the problem progresses eg. may start out normal, then become elevated because of the lesion, passing back through the normal range as the horse goes into shock 

5: pulse - linear - the heart rate in beats per minute - is a reflection of the heart condition: 30 -40 is normal for adults - rare to have a lower than normal rate although athletic horses may have a rate of 20-25 - animals with painful lesions or suffering from circulatory shock may have an elevated heart rate

6: respiratory rate - linear - normal rate is 8 to 10 - usefulness is doubtful due to the great fluctuations

7: temperature of extremities - a subjective indication of peripheral circulation - possible values: 1 = Normal 2 = Warm 3 = Cool 4 = Cold - cool to cold extremities indicate possible shock - hot extremities should correlate with an elevated rectal temp. 

11: pain - a subjective judgement of the horse's pain level - possible values: 1 = alert, no pain 2 = depressed 3 = intermittent mild pain 4 = intermittent severe pain 5 = continuous severe pain - should NOT be treated as a ordered or discrete variable! - In general, the more painful, the more likely it is to require surgery - prior treatment of pain may mask the pain level to some extent 

23: outcome - what eventually happened to the horse? - possible values: 1 = lived 2 = died 3 = was euthanized

In [6]:
sick_horse.loc[sick_horse['age'] > 2]

surgery  age  rectal temperature  pulse  respiratory rate  \
2        1.0    9                39.1  164.0              84.0   
8        2.0    9                38.3   90.0               NaN   
12       2.0    9                38.0   92.0              28.0   
15       1.0    9                 NaN  128.0              36.0   
22       1.0    9                38.3  130.0              60.0   
38       1.0    9                39.2  146.0              96.0   
40       2.0    9                39.0  150.0              72.0   
54       1.0    9                38.6  160.0              20.0   
73       1.0    9                 NaN    NaN               NaN   
74       1.0    9                39.7  100.0               NaN   
81       1.0    9                38.1  100.0              80.0   
102      1.0    9                38.0  140.0              68.0   
108      1.0    9                38.4   84.0              40.0   
134      1.0    9                38.1  136.0              48.0   
190      2.0    9                 NaN  100.0              44.0   
211      1.0    9                38.9  120.0              30.0   
226      1.0    9                 NaN    NaN               NaN   
228      1.0    9                38.5  120.0              70.0   
243      1.0    9                38.2  124.0              88.0   
251      1.0    9                38.8  124.0              36.0   
254      1.0    9                38.8  184.0              84.0   
261      2.0    9                38.6   88.0              28.0   
274      1.0    9                38.8  150.0              50.0   
286      2.0    9                39.5   84.0              30.0   

     temperature of extremities  pain  outcome  
2                           4.0   2.0      2.0  
8                           1.0   5.0      1.0  
12                          1.0   1.0      2.0  
15                          3.0   4.0      2.0  
22                          NaN   2.0      1.0  
38                          NaN   NaN      2.0  
40                          NaN   NaN      1.0  
54                          3.0   3.0      2.0  
73                          NaN   NaN      2.0  
74                          3.0   2.0      3.0  
81                          3.0   3.0      1.0  
102                         1.0   3.0      1.0  
108                         3.0   3.0      2.0  
134                         3.0   5.0      2.0  
190                         2.0   4.0      1.0  
211                         1.0   3.0      1.0  
226                         NaN   NaN      1.0  
228                         NaN   NaN      1.0  
243                         1.0   2.0      1.0  
251                         3.0   2.0      2.0  
254                         1.0   4.0      2.0  
261                         NaN   NaN      1.0  
274                         1.0   5.0      2.0  
286                         NaN   NaN      1.0

In [7]:
# заменим эти значения на 2, т.к в описание нет 9
sick_horse['age'] = sick_horse['age'].replace(9, 2)
sick_horse.head(5)

surgery  age  rectal temperature  pulse  respiratory rate  \
0      1.0    1                39.2   88.0              20.0   
1      2.0    1                38.3   40.0              24.0   
2      1.0    2                39.1  164.0              84.0   
3      2.0    1                37.3  104.0              35.0   
4      2.0    1                 NaN    NaN               NaN   

   temperature of extremities  pain  outcome  
0                         NaN   3.0      3.0  
1                         1.0   3.0      1.0  
2                         4.0   2.0      2.0  
3                         NaN   NaN      2.0  
4                         2.0   2.0      1.0

### Задание 2. Первичное изучение данных

In [8]:
# поиск межквартильного размаха по пульсу
q1 = sick_horse['pulse'].quantile(0.25)
q3 = sick_horse['pulse'].quantile(0.75)
IQR = q3 - q1
lower_bound = q1 - (1.5 * IQR) 
upper_bound = q3 + (1.5 * IQR)
remove_outliers = sick_horse[sick_horse['pulse'].between(lower_bound, upper_bound, inclusive=True)]
# вывод выбросов
pd.concat([sick_horse, remove_outliers]).drop_duplicates(keep=False).sort_values('pulse')
# в норме 30 - 40 ударов в минуту, у этих лошадей пульс значительно превышен, возможно они больны

C:\Users\myaut\AppData\Local\Temp\ipykernel_35228\4108944057.py:7: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  remove_outliers = sick_horse[sick_horse['pulse'].between(lower_bound, upper_bound, inclusive=True)]


surgery  age  rectal temperature  pulse  respiratory rate  \
40       2.0    2                39.0  150.0              72.0   
274      1.0    2                38.8  150.0              50.0   
54       1.0    2                38.6  160.0              20.0   
2        1.0    2                39.1  164.0              84.0   
254      1.0    2                38.8  184.0              84.0   
4        2.0    1                 NaN    NaN               NaN   
51       2.0    1                 NaN    NaN               NaN   
57       1.0    1                 NaN    NaN              20.0   
73       1.0    2                 NaN    NaN               NaN   
92       2.0    1                 NaN    NaN               NaN   
114      2.0    1                 NaN    NaN              40.0   
116      2.0    1                39.5    NaN               NaN   
125      1.0    1                38.7    NaN               NaN   
150      1.0    1                39.3    NaN               NaN   
158      2.0    1                 NaN    NaN              12.0   
159      1.0    1                 NaN    NaN               NaN   
173      2.0    1                 NaN    NaN               NaN   
203      1.0    1                39.2    NaN              23.0   
215      2.0    1                 NaN    NaN               NaN   
226      1.0    2                 NaN    NaN               NaN   

     temperature of extremities  pain  outcome  
40                          NaN   NaN      1.0  
274                         1.0   5.0      2.0  
54                          3.0   3.0      2.0  
2                           4.0   2.0      2.0  
254                         1.0   4.0      2.0  
4                           2.0   2.0      1.0  
51                          1.0   1.0      1.0  
57                          4.0   5.0      2.0  
73                          NaN   NaN      2.0  
92                          3.0   5.0      2.0  
114                         3.0   3.0      1.0  
116                         3.0   3.0      3.0  
125                         3.0   5.0      1.0  
150                         4.0   4.0      2.0  
158                         1.0   2.0      1.0  
159                         NaN   4.0      1.0  
173                         3.0   4.0      2.0  
203                         3.0   4.0      1.0  
215                         NaN   NaN      1.0  
226                         NaN   NaN      1.0

In [9]:
# поиск межквартильного размаха по частоте дыхания
q1 = sick_horse['respiratory rate'].quantile(0.25)
q3 = sick_horse['respiratory rate'].quantile(0.75)
IQR = q3 - q1
lower_bound = q1 - (1.5 * IQR) 
upper_bound = q3 + (1.5 * IQR)
remove_outliers = sick_horse[sick_horse['respiratory rate'].between(lower_bound, upper_bound, inclusive=True)]
# вывод выбросов
pd.concat([sick_horse, remove_outliers]).drop_duplicates(keep=False).sort_values('respiratory rate')
# в норме 8-16 вдохов и выдохов в минуту, лошадкам явно нехорошо 

# оставляю выбросы и продолжу работу с ними т.к. судя по пульсу, частоте дыхания - это естественные выбросы(лошади болееют)

C:\Users\myaut\AppData\Local\Temp\ipykernel_35228\3595104263.py:7: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  remove_outliers = sick_horse[sick_horse['respiratory rate'].between(lower_bound, upper_bound, inclusive=True)]


surgery  age  rectal temperature  pulse  respiratory rate  \
119      1.0    1                39.4   54.0              66.0   
102      1.0    2                38.0  140.0              68.0   
124      1.0    1                38.0   42.0              68.0   
264      2.0    1                 NaN   56.0              68.0   
228      1.0    2                38.5  120.0              70.0   
..       ...  ...                 ...    ...               ...   
255      1.0    1                37.5   72.0               NaN   
270      1.0    1                37.7   80.0               NaN   
273      1.0    1                 NaN   76.0               NaN   
280      2.0    1                40.0   78.0               NaN   
283      2.0    1                38.5   54.0               NaN   

     temperature of extremities  pain  outcome  
119                         1.0   2.0      1.0  
102                         1.0   3.0      1.0  
124                         4.0   3.0      1.0  
264                         3.0   3.0      3.0  
228                         NaN   NaN      1.0  
..                          ...   ...      ...  
255                         2.0   2.0      3.0  
270                         3.0   5.0      1.0  
273                         NaN   NaN      3.0  
280                         3.0   2.0      2.0  
283                         1.0   3.0      1.0  

[62 rows x 8 columns]

In [10]:
# посмотрим на возраст и сравним их
sick_horse['age'].value_counts()

1    275
2     24
Name: age, dtype: int64

In [11]:
young_horse = sick_horse[sick_horse['age'] == 1]
young_horse

surgery  age  rectal temperature  pulse  respiratory rate  \
0        1.0    1                39.2   88.0              20.0   
1        2.0    1                38.3   40.0              24.0   
3        2.0    1                37.3  104.0              35.0   
4        2.0    1                 NaN    NaN               NaN   
5        1.0    1                37.9   48.0              16.0   
..       ...  ...                 ...    ...               ...   
294      1.0    1                 NaN  120.0              70.0   
295      2.0    1                37.2   72.0              24.0   
296      1.0    1                37.5   72.0              30.0   
297      1.0    1                36.5  100.0              24.0   
298      1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           NaN   3.0      3.0  
1                           1.0   3.0      1.0  
3                           NaN   NaN      2.0  
4                           2.0   2.0      1.0  
5                           1.0   3.0      1.0  
..                          ...   ...      ...  
294                         4.0   2.0      3.0  
295                         3.0   4.0      3.0  
296                         4.0   4.0      2.0  
297                         3.0   3.0      1.0  
298                         NaN   NaN      3.0  

[275 rows x 8 columns]

In [12]:
adult_horse = sick_horse[sick_horse['age'] == 2]
adult_horse

surgery  age  rectal temperature  pulse  respiratory rate  \
2        1.0    2                39.1  164.0              84.0   
8        2.0    2                38.3   90.0               NaN   
12       2.0    2                38.0   92.0              28.0   
15       1.0    2                 NaN  128.0              36.0   
22       1.0    2                38.3  130.0              60.0   
38       1.0    2                39.2  146.0              96.0   
40       2.0    2                39.0  150.0              72.0   
54       1.0    2                38.6  160.0              20.0   
73       1.0    2                 NaN    NaN               NaN   
74       1.0    2                39.7  100.0               NaN   
81       1.0    2                38.1  100.0              80.0   
102      1.0    2                38.0  140.0              68.0   
108      1.0    2                38.4   84.0              40.0   
134      1.0    2                38.1  136.0              48.0   
190      2.0    2                 NaN  100.0              44.0   
211      1.0    2                38.9  120.0              30.0   
226      1.0    2                 NaN    NaN               NaN   
228      1.0    2                38.5  120.0              70.0   
243      1.0    2                38.2  124.0              88.0   
251      1.0    2                38.8  124.0              36.0   
254      1.0    2                38.8  184.0              84.0   
261      2.0    2                38.6   88.0              28.0   
274      1.0    2                38.8  150.0              50.0   
286      2.0    2                39.5   84.0              30.0   

     temperature of extremities  pain  outcome  
2                           4.0   2.0      2.0  
8                           1.0   5.0      1.0  
12                          1.0   1.0      2.0  
15                          3.0   4.0      2.0  
22                          NaN   2.0      1.0  
38                          NaN   NaN      2.0  
40                          NaN   NaN      1.0  
54                          3.0   3.0      2.0  
73                          NaN   NaN      2.0  
74                          3.0   2.0      3.0  
81                          3.0   3.0      1.0  
102                         1.0   3.0      1.0  
108                         3.0   3.0      2.0  
134                         3.0   5.0      2.0  
190                         2.0   4.0      1.0  
211                         1.0   3.0      1.0  
226                         NaN   NaN      1.0  
228                         NaN   NaN      1.0  
243                         1.0   2.0      1.0  
251                         3.0   2.0      2.0  
254                         1.0   4.0      2.0  
261                         NaN   NaN      1.0  
274                         1.0   5.0      2.0  
286                         NaN   NaN      1.0

In [13]:
print(f'Молодые лошади\n{young_horse.mean()}')
print('-' * 50)
print(f'Взрослые лошади\n{adult_horse.mean()}')

Молодые лошади
surgery                        1.408759
age                            1.000000
rectal temperature            38.122831
pulse                         67.462451
respiratory rate              28.239819
temperature of extremities     2.361233
pain                           2.929515
outcome                        1.551095
dtype: float64
--------------------------------------------------
Взрослые лошади
surgery                         1.250000
age                             2.000000
rectal temperature             38.645000
pulse                         123.363636
respiratory rate               54.600000
temperature of extremities      2.125000
pain                            3.117647
outcome                         1.541667
dtype: float64


In [14]:
# у взрослыч лошадей выше температура, учащен пульс и дыхание, испытывают легкую боль, у молодых - подавленое настроение

In [15]:
# посмотрим на оперированых лошадей и тех кого, лечили консервативно
surgery_horse = sick_horse[sick_horse['surgery'] == 1]
surgery_horse

surgery  age  rectal temperature  pulse  respiratory rate  \
0        1.0    1                39.2   88.0              20.0   
2        1.0    2                39.1  164.0              84.0   
5        1.0    1                37.9   48.0              16.0   
6        1.0    1                 NaN   60.0               NaN   
9        1.0    1                38.1   66.0              12.0   
..       ...  ...                 ...    ...               ...   
292      1.0    1                 NaN   78.0              24.0   
294      1.0    1                 NaN  120.0              70.0   
296      1.0    1                37.5   72.0              30.0   
297      1.0    1                36.5  100.0              24.0   
298      1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           NaN   3.0      3.0  
2                           4.0   2.0      2.0  
5                           1.0   3.0      1.0  
6                           3.0   NaN      2.0  
9                           3.0   3.0      1.0  
..                          ...   ...      ...  
292                         3.0   NaN      3.0  
294                         4.0   2.0      3.0  
296                         4.0   4.0      2.0  
297                         3.0   3.0      1.0  
298                         NaN   NaN      3.0  

[180 rows x 8 columns]

In [16]:
no_sergery_horse = sick_horse[sick_horse['surgery'] == 2]
no_sergery_horse

surgery  age  rectal temperature  pulse  respiratory rate  \
1        2.0    1                38.3   40.0              24.0   
3        2.0    1                37.3  104.0              35.0   
4        2.0    1                 NaN    NaN               NaN   
7        2.0    1                 NaN   80.0              36.0   
8        2.0    2                38.3   90.0               NaN   
..       ...  ...                 ...    ...               ...   
285      2.0    1                37.8   82.0              12.0   
286      2.0    2                39.5   84.0              30.0   
289      2.0    1                38.6   45.0              16.0   
293      2.0    1                38.5   40.0              16.0   
295      2.0    1                37.2   72.0              24.0   

     temperature of extremities  pain  outcome  
1                           1.0   3.0      1.0  
3                           NaN   NaN      2.0  
4                           2.0   2.0      1.0  
7                           3.0   4.0      3.0  
8                           1.0   5.0      1.0  
..                          ...   ...      ...  
285                         3.0   4.0      3.0  
286                         NaN   NaN      1.0  
289                         2.0   1.0      1.0  
293                         1.0   2.0      1.0  
295                         3.0   4.0      3.0  

[118 rows x 8 columns]

In [17]:
print(f'Молодые лошади\n{surgery_horse.mean()}')
print('-' * 50)
print(f'Взрослые лошади\n{no_sergery_horse.mean()}')

Молодые лошади
surgery                        1.000000
age                            1.100000
rectal temperature            38.143571
pulse                         76.396341
respiratory rate              33.375000
temperature of extremities     2.450000
pain                           3.293706
outcome                        1.622222
dtype: float64
--------------------------------------------------
Взрослые лошади
surgery                        2.000000
age                            1.050847
rectal temperature            38.201020
pulse                         65.500000
respiratory rate              26.114583
temperature of extremities     2.196078
pain                           2.430000
outcome                        1.440678
dtype: float64


### Задание 3. Работа с пропусками

In [18]:
# общее кол-во пропусков %
(sick_horse.isna().mean() * 100).round(2)

surgery                        0.33
age                            0.00
rectal temperature            20.07
pulse                          8.03
respiratory rate              19.40
temperature of extremities    18.73
pain                          18.39
outcome                        0.33
dtype: float64

In [19]:
# т.к у нас отсутствуют столбцы с пропусками более 80%, то все столбцы оставляем

In [20]:
print(sick_horse['temperature of extremities'].value_counts())
print('-' * 30)
print(sick_horse['pain'].value_counts())
print('-' * 30)
print(sick_horse['outcome'].value_counts())

3.0    108
1.0     78
2.0     30
4.0     27
Name: temperature of extremities, dtype: int64
------------------------------
3.0    67
2.0    59
5.0    41
4.0    39
1.0    38
Name: pain, dtype: int64
------------------------------
1.0    178
2.0     76
3.0     44
Name: outcome, dtype: int64


In [21]:
# Гипотеза - проверим взаимосвязь между столбцами pain и outcome с группировкой по возрасту лошади
print(sick_horse.groupby(['age','pain'])['outcome'].agg(pd.Series.mode))

age  pain
1    1.0            1.0
     2.0            1.0
     3.0            1.0
     4.0            2.0
     5.0            2.0
2    1.0            2.0
     2.0     [1.0, 2.0]
     3.0            1.0
     4.0            2.0
     5.0            2.0
Name: outcome, dtype: object


In [22]:
# можно сделать вывод, что молодые лошади при коликах практически всегда погибают, (исключение тольуо уровень боли 3).
# Взрослые лошади при низком уровне боли (до 4) остются живыми, а с 4 уровня боли погибают все лошади. 
# Признаем эту гипотезу допустимой. Поэтому недостающие значения в столбце outcome заполняеем исходя из значения столбца pain

In [23]:
#Сначала заполним недостающими значениями столбец outcome (у него самый маленький процент пропусков)
#Все действия будем производить с новым датафреймом, для быстрого отката в случае неправильного применения фильтра
#Заполним недостающие значениями столбца outcome

fill_horses_df = sick_horse
fill_horses_df['outcome'].fillna(fill_horses_df.groupby(['pain','age'])['outcome'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  298.000000  299.000000          239.000000  275.000000   
mean     1.395973    1.080268           38.166527   71.934545   
std      0.489881    0.272162            0.733508   28.680522   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        241.000000                  243.000000  244.000000  299.000000  
mean          30.427386                    2.345679    2.942623    1.551839  
std           17.678256                    1.046369    1.303993    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [24]:
# Гипотеза - проверим взаимосвязь между столбцами surgery?(операция) и pain (боль) (тем более это описано в документации)
print(sick_horse.groupby(['surgery'])['pain'].median())

surgery
1.0    3.0
2.0    2.0
Name: pain, dtype: float64


In [25]:
# взаимосвязь есть - операция была (1) при высоком медианном значении боли(3), и не было(2) при низком значении боли(2)

In [26]:
#заполним недостающие значениями столбца pain исходя из медианных значений столбца surgery
fill_horses_df['pain'].fillna(sick_horse.groupby(['surgery'])['pain'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  298.000000  299.000000          239.000000  275.000000   
mean     1.395973    1.080268           38.166527   71.934545   
std      0.489881    0.272162            0.733508   28.680522   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        241.000000                  243.000000  299.000000  299.000000  
mean          30.427386                    2.345679    2.892977    1.551839  
std           17.678256                    1.046369    1.199238    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [27]:
#  в подтверждение гипотезы роверим обратную взаимосвязь между столбцами pain(боль) и surgery(хтрургическое вмешательство)
print(sick_horse.groupby(['pain'])['surgery'].median())

pain
1.0    2.0
2.0    2.0
3.0    1.0
4.0    1.0
5.0    1.0
Name: surgery, dtype: float64


In [28]:
# взаимосвязь есть - что на уровнях боли 3 и выше всегда было хирургическое вмешательство. Восстановим отсутствующее значение в столбце 'surgery'

In [29]:
fill_horses_df['surgery'].fillna(fill_horses_df.groupby(['pain'])['surgery'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  299.000000          239.000000  275.000000   
mean     1.394649    1.080268           38.166527   71.934545   
std      0.489595    0.272162            0.733508   28.680522   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        241.000000                  243.000000  299.000000  299.000000  
mean          30.427386                    2.345679    2.892977    1.551839  
std           17.678256                    1.046369    1.199238    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [30]:
# Гипотеза - проверим взаимосвязь между outcome  и temperature of extremities
print(sick_horse.groupby(['outcome'])['temperature of extremities'].median())

outcome
1.0    2.0
2.0    3.0
3.0    3.0
Name: temperature of extremities, dtype: float64


In [31]:
# Вывод - выявлена прямая связь-в случаях смертности (outcome =2 или 3),температура экскрементов (temperature of extremities) всегда была 3(это признак шока - из описания). 
# В случае если outcome = 1 (лошадь осталась живой), температура экскрементов = 2 (норма)

In [32]:
#Восстановим отсутствующие значения в столбце 'temperature of extremities' 

fill_horses_df['temperature of extremities'].fillna(fill_horses_df.groupby(['outcome'])['temperature of extremities'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  299.000000          239.000000  275.000000   
mean     1.394649    1.080268           38.166527   71.934545   
std      0.489595    0.272162            0.733508   28.680522   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        241.000000                  299.000000  299.000000  299.000000  
mean          30.427386                    2.361204    2.892977    1.551839  
std           17.678256                    0.967578    1.199238    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [33]:
# Гипотеза- проверим взаимосвязь между столбцами temperature of extremities и pulse
print(sick_horse.groupby(['temperature of extremities'])['pulse'].median())

temperature of extremities
1.0    54.0
2.0    54.0
3.0    78.0
4.0    90.0
Name: pulse, dtype: float64


In [34]:
# Вывод- выявлена прямая связь - чем ниже температура экскрементов(признак заболевания) - тем выше пульс

In [35]:
#Восстановим отсутствующие значения в столбце 'pulse' 

fill_horses_df['pulse'].fillna(fill_horses_df.groupby(['temperature of extremities'])['pulse'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  299.000000          239.000000  299.000000   
mean     1.394649    1.080268           38.166527   71.779264   
std      0.489595    0.272162            0.733508   27.745453   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        241.000000                  299.000000  299.000000  299.000000  
mean          30.427386                    2.361204    2.892977    1.551839  
std           17.678256                    0.967578    1.199238    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [36]:
#Гипотеза - проверим утверждения из описания, что ректальная температура повышается во время инфекции,а во время болевого 
#шока(значение-pain) температура понижается
print(sick_horse.groupby(['pain'])['rectal temperature'].median())

pain
1.0    38.2
2.0    38.3
3.0    38.2
4.0    38.0
5.0    38.0
Name: rectal temperature, dtype: float64


In [37]:
# Вывод- гипотезу можно признать успешной. При высоких показателях боли медианная температура понижается.

In [38]:
#Восстановим отсутствующие значения в столбце 'rectal temperature' 

fill_horses_df['rectal temperature'].fillna(fill_horses_df.groupby(['pain'])['rectal temperature'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  299.000000          299.000000  299.000000   
mean     1.394649    1.080268           38.158194   71.779264   
std      0.489595    0.272162            0.658164   27.745453   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        241.000000                  299.000000  299.000000  299.000000  
mean          30.427386                    2.361204    2.892977    1.551839  
std           17.678256                    0.967578    1.199238    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [39]:
# Остался последний показатель- respiratory rate. В описании так и написано- полезность сомнительная из-за больших колебаний. Поэтому заполним недостающие значения медианными.

In [40]:
fill_horses_df['respiratory rate'].fillna(fill_horses_df['respiratory rate'].median(), inplace=True)
fill_horses_df.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  299.000000          299.000000  299.000000   
mean     1.394649    1.080268           38.158194   71.779264   
std      0.489595    0.272162            0.658164   27.745453   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        299.000000                  299.000000  299.000000  299.000000  
mean          29.180602                    2.361204    2.892977    1.551839  
std           16.067835                    0.967578    1.199238    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           20.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           34.500000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [41]:
fill_horses_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   age                         299 non-null    int64  
 2   rectal temperature          299 non-null    float64
 3   pulse                       299 non-null    float64
 4   respiratory rate            299 non-null    float64
 5   temperature of extremities  299 non-null    float64
 6   pain                        299 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.8 KB
